src/main/java/net/neoforged/neoforge/transfer/handlers/templates/resource/IResourceStorageData.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.handlers.templates.resource;
+
+import java.util.Iterator;
+import net.neoforged.neoforge.transfer.resources.IResource;
+import net.neoforged.neoforge.transfer.resources.IResourceStack;
+import org.jetbrains.annotations.Contract;
+
+public interface IResourceStorageData<R extends IResource> extends Iterable<IResourceStack<R>> {
+    int size();
+
+    IResourceStack<R> get(int index);
+
+    IResourceStorageData<R> modify(int index, R resource, int amount);
+
+    ResourceStorageComponent<R> component();
```

Comment:
A component here doesn't make sense since component-backed handlers will typically require a separate implementation due to using the context.